document how `--gc:arc` and `--gc:orc` work

zerbina opened the pull request:
## Summary

Only a very short overview of both was previously provided, with the
details only available by inspecting the implementation in the compiler
and its runtime.

In order to make discussion surrounding both easier and more accessible,
a detailed description of how both ARC and ORC work is now provided in
`gc.rst`.

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

## Notes for Reviewers
* suggestions for wording improvement and fixes for errors are welcome
* feel free to also state what kind of information you think is missing

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+The algorithm
+-------------
+
+The cycle colletor used for ORC is based on "trial deletion". As the first
```
```suggestion
The cycle collector used for ORC is based on "trial deletion". As the first
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+up. Clean up happens in two steps: first, all outgoing edges relevant to the
+cycle collector are physically removed (by setting the `ref` values to `nil`).
+Then, the cell is disposed by calling its destructor (if one exists) an
+freeing the underlying memory location.
```
```suggestion
Finally, all cells of the input sub-graph that are either *white* or *gray* are
cleaned up in a two step process: first, all outgoing edges relevant to the
cycle collector are physically removed (by setting the `ref` values to `nil`).
Then, the cell is disposed by calling its destructor (if one exists) an freeing
the underlying memory location.
```

Not sure if that's still accurate, but it reads a bit better. I dropped the clarification for the gray cells as the note and paragraph prior cover that.

zerbina replied:
I think it's mostly still accurate, although only mentioning a single sub-graph might be a possible source of confusion. The original paragraph was not entirely correct, however, so I rewrote it as a whole.

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+
+The implementation
+------------------
+
```
```suggestion

Collection is implemented with a 2-pass graph colouring algorithm for detection,
and a 2-step memory collection process.

```

This intro might help setup the mental model described in the paragraphs that follow.

zerbina replied:
Yep, that's better. Thanks.

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+is possible, the others are ignored and never touched.
+
+During the first pass, the sub-graph of each cell in the input set is traversed
+and each reachable *black* node (i.e. cell) as *gray* and temporarily deleting
```
```suggestion
and each reachable *black* node (i.e. cell) set to *gray* and temporarily deleting
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+
+During the first pass, the sub-graph of each cell in the input set is traversed
+and each reachable *black* node (i.e. cell) as *gray* and temporarily deleting
+its edges (i.e. decrementing the refcount of the referenced cell by one).
```
```suggestion
its edges (i.e. decrementing the refcount of each referenced cell by one).
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+
+When an edge through which a reference cycle can happen is removed (i.e. a
+`ref` is destroyed), the cell is remembered as a potential *cycle root*. Doing
+this on edge removal instad of creation has the benefit that sub-graphs that
```
```suggestion
this on edge removal instead of creation has the benefit that sub-graphs that
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+Gathering edges
+~~~~~~~~~~~~~~~
+
+To know about the outgoing edges of a cell, the cell's attached `=trace` hooks is
```
```suggestion
To know about the outgoing edges of a cell, the cell's attached `=trace` hook is
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+invoked. It is responsible for collecting all directly reachable relevant
+`ref`s of the cell to a list provided by the collector.
+
+For more information, see the `documentation <destructors.html#lifetimeminustracking-hooks-nimeqtrace-hook>`_
+of the `=trace` hook.
```
```suggestion
For more information, see the `=trace` hook `documentation 
<destructors.html#lifetimeminustracking-hooks-nimeqtrace-hook>`_.
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+For more information, see the `documentation <destructors.html#lifetimeminustracking-hooks-nimeqtrace-hook>`_
+of the `=trace` hook.
+
+Thread safety
```
```suggestion
Thread Safety
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+the threshold is reset to the default value -- otherwise it's increased until
+an implementation defined upper bound is reached.
+
+Gathering edges
```
```suggestion
Gathering Edges
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+`GC_partialCollect`), only the roots with an index greater than or equal to
+the speicifed `limit` are considered.
+
+Potential cycle roots
```
```suggestion
Potential Cycle Roots
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+Then, the cell is disposed by calling its destructor (if one exists) an
+freeing the underlying memory location.
+
+Input set
```
```suggestion
Input Set
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+deletion. Once done, all nodes that have no incoming edges are known to only
+be alive because of a reference cycle, and can now be freed.
+
+The implementation
```
```suggestion
The Implementation
```

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+and a `ref` as an *edge* in a directed cyclic graph. The cycle collector is
+responsible for freeing cells only referenced as part of references cycles.
+
+The algorithm
```
```suggestion
The Algorithm
```

saem reviewed (commented):
I did a first pass.

Except for typos, the suggestions I made are partially me checking my own understanding. I'll do another pass shortly.

Most minor thing is the changes to title casing, not a must have, but a direction I'd like to nudge the documetnation in over time.

saem commented on doc/gc.rst:
```diff
@@ -75,6 +75,208 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+a location allocated via either `new` or by using the object construction syntax
+where the type is a `ref T`) also stores a reference counter.
+
+When a managed cell is allocated, its counter starts at `1` (because the
+allocation created a single `ref` handle). Creating a `ref` to the cell
+increments the counter, and destroying one decrements it It is incremented when
+a new `ref` to the cell is created, and decremented when a `ref` to it is
+destroyed.
```
```suggestion
When a managed cell is allocated, its counter starts at `1` (because the
allocation created a single `ref` handle). Creating a `ref` to the cell
increments the counter, and destroying one decrements it.
```

saem reviewed (approved):
I like it!

Let's merge!

zerbina committed:
document how `--gc:arc` and `--gc:orc` work

Summary
=======

Only a very short overview of both was previously provided, with the
details only available by inspecting the implementation in the compiler
and compiler run-time.

In order to make discussion surrounding both easier and more accessible,
a detailed description of how both ARC and ORC work is now provided in
`gc.rst`.

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

saem reviewed (approved):
bors r+

zerbina commented:
bors r+
